In [1]:
import pandas as pd
import numpy as np

In [2]:
import sklearn.discriminant_analysis
sklearn.discriminant_analysis.__file__

'/Users/wendellwang/Developer/ML/lib/python3.9/site-packages/sklearn/discriminant_analysis.py'

In [3]:
db = pd.read_csv('insect.txt', 
                 header=None, 
                 names=['Y', 'X1', 'X2', 'X3'], 
                 sep=' ')
db.head()

Y   X1   X2  X3
0  a  191  131  53
1  a  185  134  50
2  a  200  137  52
3  a  173  127  50
4  a  171  128  49

In [12]:
from scipy.stats import bartlett

In [134]:
stat, p = bartlett(db['X1'], db['X2'], db['X3'])

In [135]:
p

1.0685193072965499e-12

In [136]:
stat

55.12949450156246

In [137]:
[np.var(x, ddof=1) for x in [db['X1'], db['X2'], db['X3']]]

[441.9236842105263, 85.83157894736841, 7.694736842105264]

In [157]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

In [158]:
model = LDA(store_covariance=True)

In [159]:
Y = np.array(db['Y'])
X = np.array(db[['X1', 'X2', 'X3']])

In [160]:
model.fit(X,Y)

LinearDiscriminantAnalysis(store_covariance=True)

In [161]:
print(model.predict([[191, 131, 53]]))

['a']


**Due to the limitations of sklearn, we were not able to get the linear discriminant functions for the two species separately**. We could only get the coefficients and intercept of the final result $(\hat{d_b^L}(x)-\hat{d_a^L}(x))$

In [165]:
model.coef_

array([[ 0.67922537, -0.40785906, -2.70428038]])

In [166]:
model.intercept_

array([54.09784031])

So, as the code above shown, the linear discriminant function of $(\hat{d_b^L}(x)-\hat{d_a^L}(x))$ is
$$
(\hat{d_b^L}(x)-\hat{d_a^L}(x)) = 54.098 + 0.679x_1 -0.408x_2-2.704x_3
$$
which also matches the substraction result from the website https://online.stat.psu.edu/stat505/lesson/10/10.4
$$
\hat{d_a^L}(x) = -247.276 -1.417x_1 +1.520x_2+10.954x_3
$$
$$
\hat{d_b^L}(x) = -193.178 -0.738x_1 +1.113x_2+8.250x_3
$$